### ISSUE TITLE
Failing negative assertions when a screen reader exhibits 100 percent unexpected behaviour

### ISSUE THREAD

**Issue by jscholes** [w3c/aria-at#1216]
I'm currently running through the Media Seek Slider test plan.  In a few of the tests, NVDA fails to provide any of the expected behaviour, for example:
* I'm instructed to press F to jumpt to the slider via the next form field command, but NVDA jumps to the "Run test setup" button instead.
* I'm instructed to press Insert+Up Arrow in focus mode with focus on the slider, at which point NVDA only speaks the word "blank".

In terms of positively worded assertions like "Role 'slider' is conveyed", the response is simple: If it says "blank" or moved to a button, it unambiguously does not convey the role slider.  But how should testers respond to a negatively worded assertion in this scenario?

In my Insert+Up Arrow example: I'm focused on a slider, I press it, and NVDA says "blank".  I'm then asked by the app if the following statement is true:

> Numeric value, '90', is not conveyed

To my mind, the answer to this question is a clear yes.  NVDA says "blank", not "90".  But if I select "Yes", I'm marking an assertion as passing even though NVDA fails to do what was expected of it by the test in every other aspect.

Similarly for the next form field case (command F): If NVDA announces "Run test setup button unavailable", it does not announce "90".  Should I mark this assertion as:
* "Yes"/passing, because it doesn't say "90"; or
* "No"/failing, because it fails to move to the slider at all?

In talking to @IsaDC about this, she indicated that consensus was reached in a previous Community Group meeting about how to handle these cases: If absolutely none of the expected behaviour was exhibited, we should fail all assertions.  But I'm not sure if that previous discussion took negative assertions into account.  If it did and/or still holds, this type of information isn't made available to testers upfront, requiring it to be manually communicated to all new testers in a very inefficient, error-prone way.

#### Questions
1. How should negative assertions be responded to when the technical answer is "Yes", but the screen reader failed to meet expectations to the degree that the answer is rendered essentially meaningless?
2. If the answer to `#1` is that testers should respond "No", is it a problem that we are asking people to interpret simple English phrases differently based on such vague context around how screen readers do or don't work?
3. Should the app provide an additional option, similar to the "No output" checkbox, that allows a tester to indicate that all assertions have been rendered irrelevant by a screen reader's unexpected behaviour?

Note: For now, in my test run I've marked all assertions as "No" in such cases to avoid a ton of conflicts, in line with the context that was shared with me.

CC @mcking65

**Comment by mcking65** [comment]
You raise a really good question. It is very unfortunate we have not previously addressed this in a robust manner.

I think it is important that we write tests plans in a way where we do not generate artificially correct verdicts. For example, if we have an assertion that says "do not convey value x" and the screen reader conveys X when the appropriate test conditions are present, then the correct verdict for that assertion is fail. However, if the appropriate conditions are not present and the screen reader just happens not to convey X in the inappropriate conditions, then it would be incorrect to say that the assertion passed.

In the current test plans, this means that certain assertions have implicit preconditions. For example, any time the test is to navigate to an element and then convey (or not convey) information about that element, the precondition for conveying the information is that the navigation was successful. If the navigation was not successful, then the assertions related to conveying information cannot have "pass" verdicts.

Negative assertions definitely make this problem more prominent.  In the case of the positive assertion:

"The role slider is conveyed"

It is implied that the assertion is:

"role of the element, 'slider', is conveyed."

Where "the element" means the "seek slider".

If there were two sliders and navigation went to the wrong slider, it would be incorrect to mark the role assertion as passing even if the role 'slider' was conveyed.

I don't know the best way to fix this problem without re-writing a lot of tests. I think you are on good paths with your two options:
1. A "not applicable" verdict or something of that nature.
2. A "breaking negative side effect" option where the side effect makes the test invalid.


I kind of lean toward the side effect approach because it would create a verdict with very clear meaning for both screen reader developers and web developers.

Right now, we have moderate and severe negative side effects. We automatically generate assertions for side effects. We could have a new side effect, "breaking side effect", that would automatically mark all assertions as "untestable". It would set the "severe negative side effects" assertion as failing. There would be a description of the breaking side effect.

We could do this without changes to current test plans. It would require a small adjustment to the test runner and then some adjustments to the report generation code.
